### DISCUSSION TITLE
Git Branching and Merging Strategy for a Shared Team Resource

### DISCUSSION THREAD

**Discussion by ckkoeber** [community/community#21756] [New to GitHub]
<p>Hello to All,</p>
<p>My team is facing a source control workflow challenge with Git when it comes to having a shared resource.</p>
<p>Here is the team make-up:</p>
<ul>
<li>Sprint Team <span class="hashtag">#1</span>
</li>
<li>Sprint Team <span class="hashtag">#2</span>
</li>
<li>_ <strong>Shared</strong> _ DBA across teams</li>
<li>Dev. Lead (Myself)</li>
</ul>
<p>Now, each sprint we have available the following branches:</p>
<ol>
<li>Master (Representative of Production)</li>
<li>Hotfix (Copy of master for push/pull of hotfixes)</li>
<li>Release (What will be deployed next release; reconciliation of all further branches below)</li>
<li>sprint/release team <span class="hashtag">#1</span> branch (Represents all consolidated code for team <span class="hashtag">#1</span>; changes per sprint and reconciled to release branch at end of sprint)</li>
<li>sprint/release team <span class="hashtag">#2</span> branch (Represents all consolidated code for team <span class="hashtag">#2</span>; changes per sprint and reconciled to release branch at end of sprint)</li>
<li>shared resource branch (For shared DBA)</li>
<li>Other personal branches for developers (not to be considered authoritative for reconciliation)</li>
</ol>
<p><em>See attached for diagrammed version of the branching structure above <div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/1X/ce26e8de25be259ed91eceba1810e6268b33dc37.png" data-download-href="/uploads/short-url/tpHIN1yBpDgJI47yZl2actzdnvN.png?dl=1" title="Git Source Control Problem to Post Online.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/40242566/181066557-60efd327-4a82-4eb6-bedf-1a2d57bd06df.png" alt="Git Source Control Problem to Post Online.png" data-base62-sha1="tpHIN1yBpDgJI47yZl2actzdnvN" width="644" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/1X/ce26e8de25be259ed91eceba1810e6268b33dc37_2_644x500.png, https://aws1.discourse-cdn.com/github/optimized/1X/ce26e8de25be259ed91eceba1810e6268b33dc37_2_966x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/1X/ce26e8de25be259ed91eceba1810e6268b33dc37_2_1288x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/1X/ce26e8de25be259ed91eceba1810e6268b33dc37_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Git Source Control Problem to Post Online.png</span><span class="informations">1748×1357 69 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div></em></p>
<p>The problem is that the shared DBA needs to constantly and consistently make changes <em>across</em> teams for SQL-related items. In addition, for just SQL related items his changes need to be considered authoritative. His changes are localized to a specific folder, if that helps.</p>
<p><strong>Is there either a workflow or git configuration that will allow for the shared resource to effectively work either on his own branch while allowing for the team to effectively collaborate with his changes?</strong></p>
<p>Thank you for your time.</p>

**Comment by lee-dohm** [comment, marked as answer]
<p>The basic git branching workflow is this:</p>
<ol>
<li>Pick a base branch</li>
<li>Make a new branch from the base to keep your changes on</li>
<li>Make the changes</li>
<li>Do whatever testing and review you need</li>
<li>Merge the new branch into the base branch</li>
<li>
<strong>Optionally: </strong> delete the new branch</li>
</ol>
<p>So let’s say that you have the following setup at the start of sprint S:</p>
<ul>
<li>Team 1 has branch <code>team-1-development</code> pointing to <code>current-release</code>
</li>
<li>Team 2 has branch <code>team-2-development</code> pointing to <code>current-release</code>
</li>
</ul>
<p>As the sprint continues, members of Team 1 and Team 2 follow the basic git workflow by:</p>
<ol>
<li>Pick their team’s base branch: either <code>team-1-development</code> or <code>team-2-development</code>
</li>
<li>Make a new branch from the base to keep their changes on</li>
<li>Make the changes</li>
<li>Merge the new branch back into the base branch</li>
</ol>
<p>So your shared resource has two options when doing their work:</p>
<ul>
<li>Start from a team’s base branch</li>
<li>Start from their own base branch</li>
</ul>
<p>So let’s say that in sprint S, the teams request the following:</p>
<ul>
<li>Team 1 requests the shared resource create X</li>
<li>Team 2 requests the shared resource update Y with Z</li>
</ul>
<p>If the two changes are independent, in other words they can be developed completely separately from each other, then the shared resource could choose to do the following for Team 1’s request:</p>
<ol>
<li>Pick Team 1’s base branch <code>team-1-development</code>
</li>
<li>Make a new branch from the base to keep their changes on</li>
<li>Make the changes</li>
<li>Merge the new branch back into the base branch</li>
</ol>
<p>And then do the same for Team 2’s request, just replace <code>team-1-development</code> with <code>team-2-development</code>.</p>
<p>But if the two changes are dependent, in other words something in “update Y with Z” will require the work from “create X” to be done first, then the shared resource could choose to do the following:</p>
<ol>
<li>Pick <code>current-release</code> as the base branch</li>
<li>Make a new branch from the base to keep their changes on, let’s call it <code>shared-resource-1-development</code>
</li>
<li>Make the changes to create X</li>
<li>Merge the new branch  <strong>into <code>team-1-development</code></strong>
</li>
</ol>
<p>Now you might say, “Whoa! Waitaminute! This isn’t the basic git workflow at all, Lee! What are you trying to pull?” Well, it is the basic git workflow, just modified a little. You see, back up above I said that both Team 1 and the shared resource created their branches based on <code>current-release</code>. This means that when you merge <code>shared-resource-1-development</code> into <code>team-1-development</code> that git will compare the current state of <code>shared-resource-1-development</code> and <code>team-1-development</code> to <code>current-release</code>, merge both sets of changes (relative to <code>current-release</code>), and store the result in <code>team-1-development</code>. This is called a <a href="https://stackoverflow.com/a/4130766/1954" rel="nofollow noopener">three-way merge</a>.</p>
<p>Then when the shared resource moves on to the “update Y” task, they would:</p>
<ol>
<li>Make the changes to <code>shared-resource-1-development</code> to update Y with Z</li>
<li>Merge the <code>shared-resource-1-development</code> branch into <code>team-2-development</code>
</li>
</ol>
<p>This results in another three-way merge into <code>team-2-development</code> which receives both the output of the “create X” task and the output of the “update Y” task.</p>
<p>The thing to keep in mind is that all of these are essentially the same workflow:</p>
<ol>
<li>Pick a point to start your branch</li>
<li>Create a new branch from that starting point</li>
<li>Make changes</li>
<li>Merge your new branch into whatever branch you want to receive your new changes</li>
</ol>
<p>Keep in mind that: The more your two branches have in common when merging, the easier the merge process will be.</p>
<p>I hope that helps!</p>

**Reply by JohnCitz** [reply]
Beautifully answered!  Well explained and with good, realistic examples. Cheers!

**Comment by ckkoeber** [comment]
<p>Thank you for this write-up! Reviewing but looks great so far. Will ask questions when done.</p>
